src/main/java/com/treasuredata/android/GetAdvertisingIdAsyncTask.java

```diff
@@ -0,0 +1,53 @@
+package com.treasuredata.android;
+
+import android.os.AsyncTask;
+import android.content.Context;
+
+import org.komamitsu.android.util.Log;
+
+import java.lang.reflect.Method;
+
+class GetAdvertisingIdAsyncTask extends AsyncTask<Context, Void, String> {
+    private static final String TAG = GetAdvertisingIdAsyncTask.class.getSimpleName();
+    private static Object advertisingInfo;
+    private static Method isLimitAdTrackingEnabledMethod;
+    private static Method getIdMethod;
+    private final GetAdvertisingIdAsyncTaskCallback callback;
+
+    GetAdvertisingIdAsyncTask(GetAdvertisingIdAsyncTaskCallback callback) {
+        this.callback = callback;
+    }
+
+    @Override
+    protected String doInBackground(Context... params) {
+        Context context = params[0];
+        try {
+            if (advertisingInfo == null) {
```

Comment:
First, sorry, my bad, doesn't read it thoroughly earlier to realize this only gets executed on `enableAutoAppendAdvertisingIdentifier`, please remove these caches.
(This actually doesn't what I meant too. It is expensive when the class hasn't been loaded. If this task get executed multiple times, it will still try to look up for the class no matter how many time it fails, and that's where it is expensive. Lookup for the class when it's already loaded is relatively cheap)

Reply:
I updated to synchronize cacheAdvertisingInfoClass